### PR TITLE
feature: real LN tests

### LINKED ISSUE #2650: Attach Modules using builder pattern in `gatewayd`
`gatewayd` currently attaches the modules when each client is built, which results in us needing to attach all modules (including the dummy module) even when some modules are not needed. We should switch `gatewayd` to use a builder pattern, similar to fedimintd, so that we can attach exactly the modules that are needed.

### PR DESCRIPTION
**Pull request by jkitman**
Runs our LN tests with the real nodes.

Refactors the LN tests to leverage the gateways rather than construct clients directly.  This will allow us to test gateway-federation interactions better.

Subtask  of #2315.
Fixes #2650, moving client registration into the `bin`.

### CODE REVIEW

File: gateway/ln-gateway/src/client.rs [hunk 1]
```diff
@@ -60,7 +53,8 @@ impl StandardGatewayClientBuilder {
 
         let mut client_builder = ClientBuilder::default();
         client_builder.with_module_gens(registry);
-        client_builder.with_primary_module(1);
```
Comment (jkitman) [inline gateway/ln-gateway/src/client.rs:63]: @m1sterc001guy the original code seems to set the dummy module as the primary module, no?

File: gateway/ln-gateway/src/ng/tests.rs [hunk 2]
```diff
@@ -124,7 +74,8 @@ pub fn sha256(data: &[u8]) -> sha256::Hash {
 #[tokio::test(flavor = "multi_thread")]
 async fn test_gateway_client_pay_valid_invoice() -> anyhow::Result<()> {
     gateway_test(
-        |_, other_lightning_client, _, user_client, gateway| async move {
+        |gateway, other_lightning_client, fed, user_client| async move {
+            let gateway = gateway.remove_client(&fed).await;
```
Comment (jkitman) [inline gateway/ln-gateway/src/ng/tests.rs:78]: @m1sterc001guy we remove the client from the gateway so you can manually manipulate the gateway.  However, we could also not do this and let this be more of an E2E test that will test the gateway RPC logic.

File: scripts/rust-tests.sh [hunk 3]
```diff
@@ -32,6 +32,8 @@ export FM_TEST_USE_REAL_DAEMONS=1
 
 if [ -z "${FM_TEST_ONLY:-}" ] || [ "${FM_TEST_ONLY:-}" = "bitcoind" ]; then
   >&2 echo "### Testing against bitcoind"
+  env RUST_BACKTRACE=1 cargo test -p ln-gateway -- --test-threads=$(($(nproc) * 2)) "$@"
+  env RUST_BACKTRACE=1 cargo test -p fedimint-ln-tests can_switch_active_gateway -- --test-threads=$(($(nproc) * 2)) "$@"
```
Comment (jkitman) [inline scripts/rust-tests.sh:36]: Adds all of our LN tests to run with real daemons.

File: fedimint-testing/src/fixtures.rs [hunk 4]
```diff
@@ -134,29 +126,44 @@ impl Fixtures {
         .await
     }
 
-    /// Starts a new gateway
-    pub async fn new_gateway(&self, password: Option<String>) -> GatewayTest {
+    /// Starts a new gateway with a given lightning node
+    pub async fn new_gateway(&self, ln: Arc<dyn LightningTest>) -> GatewayTest {
         // TODO: Make construction easier
```
Comment (m1sterc001guy) [inline fedimint-testing/src/fixtures.rs:131]: This probably isn't a TODO anymore right? Not sure exactly what its referring to now

File: fedimint-testing/src/gateway.rs [hunk 5]
```diff
@@ -29,56 +37,58 @@ impl GatewayTest {
...
-        fed.last().expect("No feds registered").registration.clone()
+            .unwrap()
     }
 
     pub(crate) async fn new(
         base_port: u16,
         password: String,
-        lightning: impl ILnRpcClient + 'static,
+        lightning: Arc<dyn LightningTest>,
         decoders: ModuleDecoderRegistry,
```
Comment (m1sterc001guy) [inline fedimint-testing/src/gateway.rs:58]: Can we get the decoders from the registry now that we're passing that in?

File: gateway/ln-gateway/src/bin/gatewayd.rs [hunk 6]
```diff
@@ -84,7 +85,14 @@ async fn main() -> Result<(), anyhow::Error> {
     );
 
     // Create federation client builder
-    let client_builder = StandardGatewayClientBuilder::new(data_dir.clone());
+    let mut registry = ClientModuleGenRegistry::new();
+    registry.attach(MintClientGen);
+    registry.attach(WalletClientGen::default());
```
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/bin/gatewayd.rs:90]: Step closer to this https://github.com/fedimint/fedimint/issues/2650

File: gateway/ln-gateway/src/ng/tests.rs [hunk 7]
```diff
@@ -26,93 +25,44 @@ use fedimint_ln_common::contracts::{EncryptedPreimage, FundedContract, Preimage}
...
-        },
-        timelock_delta: 10,
-        mint_channel_id: 1,
-    });
-    let gateway = fed.new_gateway_client(registry).await;
-    {
-        let fake_api = Url::from_str("http://127.0.0.1:8175").unwrap();
-        let fake_route_hints = Vec::new();
-        let register_op = gateway
-            .register_with_federation(fake_api, fake_route_hints, GW_ANNOUNCEMENT_TTL)
```
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/ng/tests.rs:70]: When does registration happen now?

Comment (m1sterc001guy) [reply gateway/ln-gateway/src/ng/tests.rs:70]: Found it, nevermind :)

File: gateway/ln-gateway/src/ng/tests.rs (hunk 2 above)
Comment (m1sterc001guy) [reply gateway/ln-gateway/src/ng/tests.rs:78]: ah ok I think I understand. So `GatewayTest` here will do actual RPCs to the webserver. I think that works, maybe we could just rename `gateway` here to be `gateway_client` to make that more clear

File: gateway/ln-gateway/tests/fixtures.rs [hunk 8]
```diff
@@ -1,39 +1,35 @@
...
-    let mut fixtures =
-        Fixtures::new_primary(1, DummyClientGen, DummyGen, DummyGenParams::default());
+    let mut fixtures = Fixtures::new_primary(DummyClientGen, DummyGen, DummyGenParams::default());
     let ln_params = LightningGenParams::regtest(fixtures.bitcoin_server());
-    fixtures = fixtures.with_module(0, LightningClientGen, LightningGen, ln_params);
+    fixtures = fixtures.with_module(LightningClientGen, LightningGen, ln_params);
 
-    let gateway = fixtures.new_gateway(password).await;
+    let lnd = fixtures.lnd().await;
+    let gateway = fixtures.new_gateway(lnd).await;
```
Comment (m1sterc001guy) [inline gateway/ln-gateway/tests/fixtures.rs:28]: I forget when this is used. Is `lnd` always the gateway when this is called?

**Review by m1sterc001guy — APPROVED** [review]

File: gateway/ln-gateway/tests/fixtures.rs (hunk 8 above)
Comment (jkitman) [reply gateway/ln-gateway/tests/fixtures.rs:28]: So `lnd()` and `cln()` will give you the two kinds of test LN nodes

You can then pass them into as many gateways as you wish.  Then connect gateways to as many feds as you wish.

File: fedimint-testing/src/gateway.rs (hunk 5 above)
Comment (justinmoon) [reply fedimint-testing/src/gateway.rs:58]: Seems like a good idea

**Pull request merged**